### PR TITLE
Fix/Tweak Smörgåsbord

### PR DESCRIPTION
**Pull request by Zhentar**
I've assembled a collection of small changes I've made locally. The common theme is 'things which seem reasonable/safe/correct even if I don't entirely remember why I did things the way I did' (or whether the things they fix can be experienced outside of other local changes).

On the ClrMd front, aside from this as far as I can tell there's the DebugModuleAndId references in WdbgExts (which can be pretty trivially converted to DEBUG_MODULE_AND_ID) and DEBUG_CREATE_PROCESS fleshed out with a few of the regular CREATE_PROCESS flags. (There's also a fix to the inappropriate `MarshalAs(UnmanagedType.LPStruct)` on GetSymbolEntryInformation but DbgShell isn't consuming it)

* Move WdbgExts into DbgProvider to reduce difference with upstream ClrMd

* The disassembly BlockId GroupBy behavior (simply print the groupby value) is a pretty sane default behavior for default GroupBys, so make it the actual default behavior.

* Pattern matching madness to simplify GroupBy code

* Let symbol name column autosize

* Memory read error handling fix/improvement

* Add HRESULT error message

* Fix address calculation for du

### CODE REVIEW

**Comment by jazzdelightsme** [comment]
Thanks for contributing! And bonus points for spelling "smörgåsbord" with the diacritics.

File: DbgProvider/internal/Util.cs [hunk 1]
```diff
@@ -416,7 +416,7 @@ public static string GetGenericTypeName( Type t, bool includeNamespace )
 
         public static string FormatErrorCode( int i )
         {
-            return FormatErrorCode( unchecked( (uint) i ) );
+            return FormatErrorCode( unchecked( (uint) i ) ) + " - " + Marshal.GetExceptionForHR( i ).Message;
```
Comment (jazzdelightsme) [inline DbgProvider/internal/Util.cs]: >" - " + Marshal.GetExceptionForHR( i ).Message; [](start = 62, length = 47)

I like the concept of this change, but I don't think this implementation will work very well--the CLR's mapping of error codes to exceptions is only a very small slice of error codes, and it's also informed by `IErrorInfo`, which when grabbing errors from a dump are not at all related.

Unfortunately I don't know of a good alternative. So for now let's just skip this particular change.

File: DbgProvider/public/Debugger/DbgEngDebugger.cs [hunk 2]
```diff
@@ -2375,6 +2357,20 @@ private void _CheckMemoryReadHr( ulong address, int hr )
             CheckHr( hr );
         } // end _CheckMemoryReadHr()
 
+        private bool _TryCheckMemoryReadHr( int hr )
+        {
+            if( (HR_ERROR_READ_FAULT == hr) ||
+                (HR_STATUS_NO_PAGEFILE == hr) ||
+                (E_NOINTERFACE == hr) ||
+                (HR_STATUS_NO_MORE_ENTRIES == hr) )
```
Comment (jazzdelightsme) [inline DbgProvider/public/Debugger/DbgEngDebugger.cs]: >HR_STATUS_NO_MORE_ENTRIES [](start = 17, length = 25)

Is this what is returned if `ReadPointersVirtual` can't read that many pointers?

File: DbgProvider/public/Debugger/DbgMemory.cs [hunk 3]
```diff
@@ -643,7 +643,7 @@ private ColorString _FormatCharsOnly( uint numColumns, Func< char, char > toDisp
                     }
 
                     cs.Append( DbgProvider.FormatAddress( addr, m_is32Bit, true, true ) ).Append( "  " );
-                    addr += (ulong) numColumns;
+                    addr += (ulong) numColumns * bytesPerChar;
```
Comment (jazzdelightsme) [inline DbgProvider/public/Debugger/DbgMemory.cs]: >bytesPerChar [](start = 49, length = 12)

D'oh! Thanks for fixing this!

File: DbgProvider/public/Debugger/DbgEngDebugger.cs (hunk 2 above)
Comment (Zhentar) [reply DbgProvider/public/Debugger/DbgEngDebugger.cs]: The only place that I remember for sure returns that is SearchVirtual; I'm not sure if there's anything else that does return it. 

It looks like ReadPointersVirtual just passes through the ReadVirtual HRESULT when it entirely fails to read a pointer, and E_FAIL when it only partially reads a pointer. (Also it looks like DbgShell's E_FAIL has a transposed 4 & 0)

Comment (jazzdelightsme) [reply DbgProvider/public/Debugger/DbgEngDebugger.cs]: Oy... looks like that `E_FAIL` is `OLE_E_NOTRUNNING`; oops. Now I wonder if I actually saw 80040005... doesn't seem likely, though; we should just fix that transposition. Thanks for finding that.

---
In reply to: [259683714](https://github.com/Microsoft/DbgShell/pull/75#discussion_r259683714) [](ancestors = 259683714)

File: DbgProvider/internal/Util.cs (hunk 1 above)
Comment (Zhentar) [reply DbgProvider/internal/Util.cs]: Alright, I took some drastic measures on this one... and read the documentation. It turns out there's an overload with a second parameter for an IErrorInfo, with `(IntPtr)(-1)` instructing it to not use any IErrorInfo at all.

Aside from that, even if the mapping isn't very complete, it does cover the ones I encounter most frequently and I think it still gives a pretty reasonable message for the remainder.

Comment (jazzdelightsme) [reply DbgProvider/internal/Util.cs]: Nice find on the -1 trick.

Even if I liked the messages from `Marshal.GetExceptionForHR`, I think I'd still want a function that just formatted the numeric error value for some situations. And I think I probably just deal with a different set of error codes than you, and always seeing "0xNNNNNNNN - Exception from HRESULT: 0xNNNNNNNN" will really annoy me. So I'd like to skip this particular change.

---
In reply to: [260593827](https://github.com/Microsoft/DbgShell/pull/75#discussion_r260593827) [](ancestors = 260593827)

Comment (Zhentar) [reply DbgProvider/internal/Util.cs]: Alright, it's removed

**Pull request merged**